### PR TITLE
Advi vignette

### PR DESCRIPTION
**Pull request by hyunjimoon**
first draft!

### CODE REVIEW

**Comment by hyunjimoon** [comment]
I added `{R  message=FALSE, warning=FALSE}` but still lengthy VI diagnostics are showing. Do you know how to make it disappear or do you recommend leaving them? [approximate_compuation.pdf](https://github.com/hyunjimoon/SBC/files/7124778/approximate_compuation.pdf)

**Comment by martinmodrak** [comment]
Hi,
looks like a good starting point, thanks!

> I added {R message=FALSE, warning=FALSE} but still lengthy VI diagnostics are showing. Do you know how to make it disappear or do you recommend leaving them

The messages are IMHO sent to default output, so you might need `results = "hide"` as well. However, the messages indicate that some fits have failed and it would be good to understand why they fail. It appears that the failure is stochastic, but it turns out one can get rid of them by adding `grad_samples = 200` to increase number of tries the system does for evaluating gradient and `init = 0.5` to avoid some rejections on initialization. I tried to summarise my experience at https://discourse.mc-stan.org/t/advi-too-many-dropped-evaluations-even-for-well-behaved-models/24338 and hope to get some additional feedback from devs who are more familiar with ADVI.

It might also be in the end necessary for us to have the ADVI backend just do multiple runs, as the algorithm appears quite unstable (although I think it would be better if this was implemented in the interfaces).



A couple additional things:

- Why is there a revert commit for "address variational backend issue 35"? I think it reintroduces some problems.
- I understand the need to reduce computation, but the low number of samples for the HMC runs is triggerring a lot of our diagnostics warnings... In fact do we even need to run HMC here? Maybe we could just show the ADVI results as we can probably easily see that the model is not broken (because it is so simple)...
- I am conflicted about the generator code - on one hand it shows a different way of generating datasets than the other vignettes, but it is IMHO much harder to understand than what we use elsewhere...
- I don't think changing `tol_rel_obj` really does anything to improve the calibration. In fact, after adding `grad_samples = 200`  it even appears to make things worse - below is my plot for default `tol_rel_obj`:

image

and this is what I get with `tol_rel_obj = 0.0001`

image

**Comment by hyunjimoon** [comment]
- Using full rank algorithm leads to better diagnostic in terms of no error (too many evaluation) and Rhat.
```
> result_0_mf <- compute_results(datasets, backend_vi_mf, thin_ranks = thin_ranks_v) 
- 25 (2%) fits resulted in an error.
 - 7 (1%) fits had at least one Rhat > 1.01. Largest Rhat was NA.
 - 5 (0%) fits had tail ESS undefined or less than half of the maximum rank, potentially skewing the rank statistics. The lowest tail ESS was NA.
 If the fits look good otherwise, increasing `thin_ranks` (via recompute_statistcs) or number of posterior samples (by refitting) might help.
Not all diagnostics are OK. You can learn more by inspecting $default_diagnostics, $backend_diagnostics and/or investigating $outputs/$messages/$warnings for detailed output from the backend.
> result_0_full <- compute_results(datasets, backend_vi_full, thin_ranks = thin_ranks_v) 
 - 8 (1%) fits had at least one Rhat > 1.01. Largest Rhat was 1.017.
Not all diagnostics are OK. You can learn more by inspecting $default_diagnostics, $backend_diagnostics and/or investigating $outputs/$messages/$warnings for detailed output from the backend.
```

File: vignettes/approximate_computation.Rmd [hunk 1]
```diff
@@ -75,85 +103,266 @@ generator_contamination <- function(N) {
...
 ds_contamination <- generate_datasets(SBC_generator_function(generator_contamination, N = 100), n_datasets = 100)
 res_contamination <- compute_results(ds_contamination, backend_contamination, cache_mode = "results", cache_location = file.path(cache_dir, "contamination"))
 ```
+The ECDF diff plot and the ranks plot look quite good, although we can notice that the `mu_signal` parameter 
+has a bit too many large ranks - the resolution we get from 100 simulations is
+however too low to be sure if that's a problem or just a fluke.
 
-```{r}
+```{r ecdf_rank_contamination}
 plot_ecdf_diff(res_contamination)
```
Comment (hyunjimoon) [inline vignettes/approximate_computation.Rmd:139]: All datasets failed to fit for the first ADVI (i.e. (Dataset i resulted in error when fitting. for all i in `res_contamination`) that line139 returns `Error: Must group by variables found in `.data`. * Column `parameter` is not found. R` which also prevents knitting.  ​Did you experience the same result? Would there be any walkaround? @martinmodrak

**Comment by hyunjimoon** [comment]
After review please merge. Thanks! @martinmodrak

File: vignettes/approximate_computation.Rmd (hunk 1 above)
Comment (martinmodrak) [reply vignettes/approximate_computation.Rmd]: I think the problem might be that you are using old version of the code. I've seen the same error after switching from a different branch. Note that the vignette uses `library(SBC)` so to have it use the development version, you need to Build+install the development version (Ctrl+Shift+B in RStudio). Once I've done that I no longer see the error.

**Review by martinmodrak — APPROVED** [review]

**Pull request merged**